Provide API context earlier to api requests in order to fix missing schedules

peppy opened the pull request:
Closes https://github.com/ppy/osu/issues/29546.

peppy committed:
Provide API context earlier to api requests in order to fix missing schedules

Closes https://github.com/ppy/osu/issues/29546.

peppy committed:
Use `IAPIProvider` interface and correctly support scheduling from `DummyAPIAccess`

peppy committed:
Fix weird test critical failure if exception happens too early in execution

Noticed in passing.

```
Exit code is 134 (Unhandled exception. System.NullReferenceException:
Object reference not set to an instance of an object.
   at osu.Game.OsuGameBase.onExceptionThrown(Exception ex) in
/Users/dean/Projects/osu/osu.Game/OsuGameBase.cs:line 695
   at osu.Framework.Platform.GameHost.abortExecutionFromException(Object
sender, Exception exception, Boolean isTerminating)
   at osu.Framework.Platform.GameHost.unobservedExceptionHandler(Object
sender, UnobservedTaskExceptionEventArgs args)
   at System.Threading.Tasks.TaskExceptionHolder.Finalize())
```

peppy committed:
Apply NRT to `APIRequest`

peppy commented on osu.Game/Online/API/APIAccess.cs:
```diff
@@ -568,6 +570,8 @@ public void Queue(APIRequest request)
         {
             lock (queue)
             {
+                request.AttachAPI(this);
```
The only important line for the fix.

bdach committed:
Fix NRT inspections

bdach reviewed (dismissed):
No objections here, if it passes tests. Maybe check on https://github.com/ppy/osu/pull/29656/commits/8ffd4aa82c5e1afd4b4fe56773d6043248b54a12 to make sure you're ok with how I fixed the inspections.

peppy commented:
Changes look fine 👍 

My rider was mid rebuilding cached so didn't show me these 😓

bdach committed:
Fix chat overlay tests

Not entirely sure why they were failing previously, but the most likely
explanation is that by freak accident some mock requests would
previously execute immediately rather than be scheduled on the API
thread, which would change execution ordering and ensure that
`ChannelManager.CurrentChannel` would become the joined channel, rather
than remaining at the channel listing.

bdach commented:
Tests did not pass but https://github.com/ppy/osu/pull/29656/commits/8b04455c29fd41dfab49974f8883acb7ef60d8fc should *probably* help. I didn't do very thorough investigating but a viable explanation is provided in the commit message.
